### PR TITLE
Self-contained deployment runtime roll forward

### LINKED ISSUE #4540: [Feature] Runtime and SDK Minor Version Roll-forward
# Runtime and SDK Minor Version Roll-forward

You will now be able to run .NET Core applications on later runtime versions, within the same major version range. For example, you will be able to run .NET Core 2.0 applications on .NET Core 2.1 or .NET Core 2.1 applications on .NET Core 2.5 (if we ever ship such a version). The roll-forward behavior is for minor versions only. 

This will cover both **Runtime** Minor Version Roll-forward, and **SDK** Minor Version Roll-forward.

### PR DESCRIPTION
**Pull request by JRAlexander**
Self-contained deployment runtime roll forward

Fixes #4540

[Internal Review URL](https://review.docs.microsoft.com/en-us/dotnet/core/deploying/runtime-patch-selection?branch=pr-en-us-5670)

### CODE REVIEW

**Review by BillWagner — APPROVED** [review]
Overall, this looks great @JRAlexander 

I had one suggestion to add, and then :shipit:

File: docs/core/deploying/runtime-patch-selection.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+[`restore`](../tools/dotnet-restore.md), [`build`](../tools/dotnet-build.md) and [`publish`](../tools/dotnet-publish.md) are `dotnet` commands that can run separately. The runtime choice is part of the `restore` operation, not `publish` or `build`. If you call `publish`, the latest patch version will be chosen. If you call `publish` with the `--no-restore` argument, then you may not get the desired patch version because a prior `restore` may not have been executed with the new self-contained application publishing policy.
```
Comment (dsplaisted) [inline docs/core/deploying/runtime-patch-selection.md]: > you may not get the desired patch version

If you would get the wrong patch version in this case, you will actually get a build error.  So I'd add the following sentence: 

> In this case, a build error will be generated with text similar to the following:
>> The project was restored using Microsoft.NETCore.App version 2.0.0, but with current settings, version 2.0.6 would be used instead. To resolve this issue, make sure the same settings are used for restore and for subsequent operations such as build or publish. Typically this issue can occur if the RuntimeIdentifier property is set during build or publish but not during restore.

Once this doc is published I think it would be a good idea to add an aka.ms link to this doc page in the error message.

File: docs/core/deploying/runtime-patch-selection.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+> [!NOTE]
+> `restore` and `build` can be run implicitly as part of another command, like `publish`. When run implicitly as part of another command, they are provided with additional context so that the right artifacts are produced. When you `publish` with a runtime (for example, `dotnet publish -r linux-x64`), the implicit `restore` restores packages for the linux-x64 runtime. If you call `restore` explicitly, it does not restore runtime packages by default, because it doesn't have that context.
+
+Running `restore` as part of the `publish` operation may be undesirable for your scenario. To avoid `restore` during `publish` while creating self-contained applications, do the following:
+
+* Set the `RuntimeIdentifiers` property to a semicolon-separated list of all the [RIDs](../rid-catalog.md) to be published
+* Set the `TargetLatestRuntimePatch` property to `true`
+
+To create both self-contained applications and [framework-dependent applications](index.md) with the same project file, use `dotnet publish --no-restore` and choose one of the following:
```
Comment (dsplaisted) [inline docs/core/deploying/runtime-patch-selection.md]: I would reword this to make this clear that this is only necessary if you don't want to restore when publishing.  If you don't need to use the `--no-restore` flag, then you don't need to do anything special in order to create both an FDD and SCD from the same project file.  You can just do, for example:

> dotnet publish
> dotnet publish -r win10-x64

Perhaps a good way to reword this would be:

>If you want to create both self-contained applications and framework-dependent applications with the same project file, and you want to use the `--no-restore` argument with `dotnet publish`, then choose one of the following:

File: docs/core/deploying/runtime-patch-selection.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+---
+title: Self-contained deployment runtime roll forward
+description: Learn about dotnet publish changes for self-contained deployments.
+author: jralexander
+ms.author: johalex
+ms.date: 05/30/2018
+---
+# Self-contained deployment runtime roll forward
+
+.NET Core [self-contained application deployments](index.md) include both the .NET Core libraries and the .NET Core runtime. Starting in .NET Core SDK 2.1.300 (.NET Core 2.1), a self-contained application deployment [now publishes the highest patch runtime on your machine](https://github.com/dotnet/designs/pull/36). By default,[`dotnet publish`](../tools/dotnet-publish.md) for a self-contained deployment selects the latest version installed as part of the SDK on the publishing machine. This enables your deployed application to run with security fixes (and other fixes) available during `publish`. The application must be re-published to obtain a new patch. Self-contained applications are created by specifying `--self-contained` on `dotnet publish` command or by specifying the RuntimeIdentifier in the project file or on the command line.
```
Comment (dsplaisted) [inline docs/core/deploying/runtime-patch-selection.md]: Normally you wouldn't explicitly specify the `--self-contained` argument.  You need to specify a RID with `-r` anyway, and if you do, then `--self-contained` will default to true.  So I'd probably replace `--self-contained` with `-r <RID>` or similar in this sentence.

**Review by dsplaisted — APPROVED** [review]

File: docs/core/deploying/runtime-patch-selection.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+> [!NOTE]
+> `restore` and `build` can be run implicitly as part of another command, like `publish`. When run implicitly as part of another command, they are provided with additional context so that the right artifacts are produced. When you `publish` with a runtime (for example, `dotnet publish -r linux-x64`), the implicit `restore` restores packages for the linux-x64 runtime. If you call `restore` explicitly, it does not restore runtime packages by default, because it doesn't have that context.
+
+Running `restore` as part of the `publish` operation may be undesirable for your scenario. To avoid `restore` during `publish` while creating self-contained applications, do the following:
+
+* Set the `RuntimeIdentifiers` property to a semicolon-separated list of all the [RIDs](../rid-catalog.md) to be published
```
[the remark this answers is not in the document]
Comment (KathleenDollard) [reply docs/core/deploying/runtime-patch-selection.md]: Yes

File: docs/core/deploying/runtime-patch-selection.md (hunk 1 above)
Comment (KathleenDollard) [reply docs/core/deploying/runtime-patch-selection.md]: Agreed

File: docs/core/deploying/runtime-patch-selection.md (hunk 2 above)
Comment (KathleenDollard) [reply docs/core/deploying/runtime-patch-selection.md]: Agreed

File: docs/core/deploying/runtime-patch-selection.md (hunk 3 above)
Comment (KathleenDollard) [reply docs/core/deploying/runtime-patch-selection.md]: Got it. Agreed

**Comment by JRAlexander** [comment]
Closing and reopening
